### DISCUSSION TITLE
function call graph which includes functions from external loaded libraries

### DISCUSSION THREAD

**Discussion by railabouni** [radareorg/radare2#21586] [Q&A]
Hi all,
I'm trying to produce a graph with all the function calls within a binary including the call graph of the used functions from external libraries.

example:
I have a binary which calls printf from libc.so.6, all the commands ("aaa"/"aaf"/"aaaa" including debugging and lazy linkage, "dcu main", "search.path /lib/aarch64-linux-gnu/libc.so.6") I tried gave me this graph or a bigger with printf as a leaf, after calling aaa and agCj:

[{"name":"entry0","size":48,"imports":["sym.imp.__libc_start_main"]},{"name":"sym.__do_global_dtors_aux","size":72,"imports":["sym.imp.__cxa_finalize","sym.deregister_tm_clones"]},{"name":"sym.print_life_rank","size":40,"imports":["sym.imp.printf"]},{"name":"sym.bar","size":48,"imports":["sym.print_life_rank","sym.imp.puts"]},{"name":"sym.foo","size":40,"imports":["sym.imp.printf"]},{"name":"main","size":140,"imports":["sym.imp.puts","sym.bar","sym.print_life_rank"]},{"name":"sym._init","size":24,"imports":["sym.call_weak_fn"]}]

when using "il" radare shows libc.so.6 as an imported library, it also shows using "dmi"

what I'm interested in is also the printf function's call grap (functions called by printf such as "__stack_chk_fail"), how can I do that?
thank you!

**Comment by trufae** [comment]
isn't agC doing what you want? do you need the json representation? see the help of ag? to render the output in the format you are interested:

```
Output formats:
| <blank>                 ascii art
| *                       r2 commands
| b                       braile art rendering (agfb)
| d                       graphviz dot
| g                       graph Modelling Language (gml)
| j                       json ('J' for formatted disassembly)
| k                       sdb key-value
| m                       mermaid
| t                       tiny ascii art
| v                       interactive ascii art
| w [path]                write to path or display graph image (see graph.gv.format)
```

**Comment by railabouni** [comment]
hello,
what I need is not a specific representation, what I need is the call graph of functions imported from shared libraries.
example: binary having function main() which calls foo() which calls bar() which calls the imported function printf()
I expect a graph like this: main()->foo()->bar()->prinft()->stack_chk_fail() because printf calls stack_chk_fail() in its implementation in libc.
what I'm getting right now is main()->foo()->bar()->prinft().

I need the call graph of the binary to show also the calls inside the imported functions from shared libs. hope this give a better explanation of what I need ... 
agC alone doesn't help

**Reply by railabouni** [reply]
here's my code if this helps with anything ... 


#include <stdio.h>

int (*fun_ptr)(int);

int foo(int life_is_good){
    printf("currently, life rank is: %d!\n",life_is_good);
    return life_is_good;
}

int print_life_rank(int life_is_good){
    printf("currently, life rank is: %d!\n",life_is_good);
    return life_is_good;
}

int bar(int life_is_good){
    print_life_rank(life_is_good);
    printf("trying to improve the world!\n");
    return life_is_good+1;
}

int main() {
    int life_is_good = -10;
   // printf() displays the string inside quotation
   while (1){
    if (life_is_good == 1){
        fun_ptr = &bar;
        printf("Life is Good! My job is done!\n");
        return 0;
    }
    else{
        life_is_good = bar(life_is_good);
        fun_ptr = &print_life_rank;
        print_life_rank(life_is_good);
    }
        (*fun_ptr)(life_is_good);
   }
   
   return 0;
}
